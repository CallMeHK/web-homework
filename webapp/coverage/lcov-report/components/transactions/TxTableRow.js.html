<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/transactions/TxTableRow.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\index.html">All files</a> / <a href="index.html">components/transactions</a> TxTableRow.js
    </h1>
    <div class='clearfix'>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from 'react'
import { transaction } from '../../utils/types'
import { css } from '@emotion/core'
import Box from '@material-ui/core/Box'
import Collapse from '@material-ui/core/Collapse'
import IconButton from '@material-ui/core/IconButton'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown'
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp'
import { TxSubTable } from './TxSubTable'
import { TxRowDropdown } from './TxRowDropdown'
import { currency } from '../../utils/currency'
&nbsp;
// const makeDataTestId = (transactionId, fieldName) =&gt; `transaction-${transactionId}-${fieldName}`
&nbsp;
export function <span class="fstat-no" title="function not covered" >TxTableRow </span>(props) {
  const { user, insertedAt, merchant, amount, credit, debit, description } = <span class="cstat-no" title="statement not covered" >props.transaction</span>
  const [open, setOpen] = <span class="cstat-no" title="statement not covered" >React.useState(false)</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >  return (</span>
    &lt;React.Fragment&gt;
      &lt;TableRow&gt;
        &lt;TableCell&gt;
          &lt;IconButton aria-label='expand row' onClick={<span class="fstat-no" title="function not covered" >()</span> =&gt; <span class="cstat-no" title="statement not covered" >setOpen(!open)}</span> size='small'&gt;
            {open ? &lt;KeyboardArrowUpIcon /&gt; : &lt;KeyboardArrowDownIcon /&gt;}
          &lt;/IconButton&gt;
        &lt;/TableCell&gt;
        &lt;TableCell&gt;{`${user.firstName} ${user.lastName}`}&lt;/TableCell&gt;
        &lt;TableCell align='right'&gt;{insertedAt}&lt;/TableCell&gt;
        &lt;TableCell align='right'&gt;{merchant.name}&lt;/TableCell&gt;
        &lt;TableCell align='right'&gt;{currency.convertToString(amount)}&lt;/TableCell&gt;
        &lt;TableCell css={css({ maxWidth: 50 })}&gt;
          &lt;TxRowDropdown transaction={props.transaction} /&gt;
        &lt;/TableCell&gt;
      &lt;/TableRow&gt;
      &lt;TableRow&gt;
        &lt;TableCell colSpan={6} style={{ paddingBottom: 0, paddingTop: 0 }}&gt;
          &lt;Collapse in={open} timeout='auto' unmountOnExit&gt;
            &lt;div css={collapsibleCellStyle}&gt;
              &lt;Box margin={1}&gt;
                &lt;TxSubTable credit={credit} debit={debit} description={description} merchant={merchant} /&gt;
              &lt;/Box&gt;
            &lt;/div&gt;
          &lt;/Collapse&gt;
        &lt;/TableCell&gt;
      &lt;/TableRow&gt;
    &lt;/React.Fragment&gt;
  )
}
&nbsp;
TxTableRow.propTypes = {
  transaction: transaction.isRequired
}
const collapsibleCellStyle = css({
  maxWidth: 600,
  width: '50%',
  minWidth: 400
})
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Jan 27 2021 01:09:09 GMT-0700 (Mountain Standard Time)
</div>
</div>
<script src="..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\sorter.js"></script>
<script src="..\..\block-navigation.js"></script>
</body>
</html>
