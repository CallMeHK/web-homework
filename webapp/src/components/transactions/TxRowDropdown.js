import React from 'react'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import Grow from '@material-ui/core/Grow'
import Paper from '@material-ui/core/Paper'
import Popper from '@material-ui/core/Popper'
import MenuItem from '@material-ui/core/MenuItem'
import MenuList from '@material-ui/core/MenuList'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import { useMutation } from '@apollo/client'
import DELETE_TRANSACTION from '../../gql/delete_transaction.gql'
import { useFormContext } from '../../context/form.context'
import { transaction } from '../../utils/types'

export const TxRowDropdown = ({ transaction }) => {
  const [open, setOpen] = React.useState(false)
  const anchorRef = React.useRef(null)
  const form = useFormContext()
  const [deleteTransaction] = useMutation(DELETE_TRANSACTION)

  const handleEdit = () => {
    form.openEditModal(transaction)
    setOpen(false)
  }

  const handleToggle = () => {
    setOpen(prevOpen => !prevOpen)
  }

  const handleClose = event => {
    if (anchorRef.current && anchorRef.current.contains(event.target)) {
      return
    }

    setOpen(false)
  }

  function handleListKeyDown(event) {
    if (event.key === 'Tab') {
      event.preventDefault()
      setOpen(false)
    }
  }

  const removeTransaction = async () => {
    await deleteTransaction({ variables: { id: transaction.id } })
    form.setData(old => {
      const transactions = old.transactions.filter(row => row.id !== transaction.id)
      return {
        ...old,
        transactions
      }
    })
  }

  const prevOpen = React.useRef(open)
  React.useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current.focus()
    }

    prevOpen.current = open
  }, [open])

  return (
    <>
      <IconButton
        aria-controls="long-menu"
        aria-haspopup="true"
        aria-label="more"
        onClick={handleToggle}
        ref={anchorRef}
        data-testid={'dropdown'}
      >
        <MoreVertIcon />
      </IconButton>
      <Popper anchorEl={anchorRef.current} open={open} role={undefined} transition>
        {({ TransitionProps, placement }) => (
          <Grow
            {...TransitionProps}
            style={{ transformOrigin: placement === 'bottom' ? 'center top' : 'center bottom' }}
          >
            <Paper>
              <ClickAwayListener onClickAway={handleClose}>
                <MenuList autoFocusItem={open} id="menu-list-grow" onKeyDown={handleListKeyDown}>
                  <MenuItem data-testid="edit" onClick={handleEdit}>
                    Edit
                  </MenuItem>
                  <MenuItem data-testid="delete" onClick={removeTransaction}>
                    Delete
                  </MenuItem>
                </MenuList>
              </ClickAwayListener>
            </Paper>
          </Grow>
        )}
      </Popper>
    </>
  )
}

TxRowDropdown.propTypes = {
  transaction: transaction.isRequired
}
