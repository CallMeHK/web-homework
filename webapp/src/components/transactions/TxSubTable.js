import React from 'react'
import { string, bool, shape } from 'prop-types'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'

export function TxSubTable ({ credit, debit, description, merchant }) {
  return (
    <TableContainer component={Paper}>
      <Table aria-label='collapsible table'>
        <TableBody>
          <TableRow>
            <TableCell>Purchase Type</TableCell>
            <TableCell data-testid="payment" align='right'>{credit ? 'Credit' : debit ? 'Debit' : 'Other'}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Transaction Description</TableCell>
            <TableCell data-testid="transactionDescription" align='right'>{description}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Merchant Description</TableCell>
            <TableCell data-testid="merchantDescription" align='right'>{merchant.description}</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableContainer>
  )
}

TxSubTable.propTypes = {
  credit: bool.isRequired,
  debit: bool.isRequired,
  description: string.isRequired,
  merchant: shape({
    description: string.isRequired
  }).isRequired
}
