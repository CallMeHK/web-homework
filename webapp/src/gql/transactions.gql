query GetDataForTable {
  transactions {
    id
    user {
      id
      firstName
      lastName
    }
    merchant {
      id
      name
      description
    }
    description
    debit
    credit
    amount
    insertedAt
  }
  merchants {
    id
    name
  }
  users {
    id
    firstName
    lastName
  }
}
