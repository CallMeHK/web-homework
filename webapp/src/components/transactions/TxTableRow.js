import React from 'react'
import { transaction } from '../../utils/types'
import { css } from '@emotion/core'
import Box from '@material-ui/core/Box'
import Collapse from '@material-ui/core/Collapse'
import IconButton from '@material-ui/core/IconButton'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown'
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp'
import { TxSubTable } from './TxSubTable'
import { TxRowDropdown } from './TxRowDropdown'
import { currency } from '../../utils/currency'

const makeDataTestId = (transactionId, fieldName) => `transaction-${transactionId}-${fieldName}`

export function TxTableRow(props) {
  const { id, user, insertedAt, merchant, amount, credit, debit, description } = props.transaction
  const [open, setOpen] = React.useState(false)

  return (
    <React.Fragment>
      <TableRow>
        <TableCell>
          <IconButton aria-label="expand row" onClick={() => setOpen(!open)} size="small">
            {open ? <KeyboardArrowUpIcon data-testid="up-icon" /> : <KeyboardArrowDownIcon data-testid="down-icon" />}
          </IconButton>
        </TableCell>
        <TableCell data-testid={makeDataTestId('userName', id)}>{`${user.firstName} ${user.lastName}`}</TableCell>
        <TableCell data-testid={makeDataTestId('transactionDate', id)} align="right">
          {insertedAt}
        </TableCell>
        <TableCell data-testid={makeDataTestId('merchantName', id)} align="right">
          {merchant.name}
        </TableCell>
        <TableCell data-testid={makeDataTestId('amount', id)} align="right">
          {currency.convertToString(amount)}
        </TableCell>
        <TableCell css={css({ maxWidth: 50 })}>
          <TxRowDropdown transaction={props.transaction} />
        </TableCell>
      </TableRow>
      <TableRow>
        <TableCell colSpan={6} style={{ paddingBottom: 0, paddingTop: 0 }}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <div css={collapsibleCellStyle}>
              <Box margin={1}>
                <TxSubTable credit={credit} debit={debit} description={description} merchant={merchant} />
              </Box>
            </div>
          </Collapse>
        </TableCell>
      </TableRow>
    </React.Fragment>
  )
}

TxTableRow.propTypes = {
  transaction: transaction.isRequired
}
const collapsibleCellStyle = css({
  maxWidth: 600,
  width: '50%',
  minWidth: 400
})
