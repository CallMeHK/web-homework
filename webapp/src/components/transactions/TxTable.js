import React from 'react'
import { arrayOf, shape, func } from 'prop-types'
import { transaction, user, merchant } from '../../utils/types'
import { css } from '@emotion/core'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import { TxTableRow } from './TxTableRow'
import { EditModal } from '../modal/EditModal'
import { FormContextProvider } from '../../context/form.context'

export function TxTable ({ data, setData }) {
  return (
    <>
      <FormContextProvider setData={setData}>
        <EditModal merchants={data.merchants} setData={setData} users={data.users} />
        <TableContainer component={Paper}>
          <Table aria-label='collapsible table'>
            <TableHead>
              <TableRow>
                <TableCell />
                <TableCell>Name</TableCell>
                <TableCell align='right'>Transaction Date</TableCell>
                <TableCell align='right'>Merchant</TableCell>
                <TableCell align='right'>Amount</TableCell>
                <TableCell css={css({ maxWidth: 50 })} />
              </TableRow>
            </TableHead>
            <TableBody>
              {data.transactions.map(row => (
                <TxTableRow key={row.id} transaction={row} />
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      </FormContextProvider>
    </>
  )
}

TxTable.propTypes = {
  data: shape({
    transactions: arrayOf(transaction),
    users: arrayOf(user),
    merchants: arrayOf(merchant)
  }).isRequired,
  setData: func.isRequired
}
