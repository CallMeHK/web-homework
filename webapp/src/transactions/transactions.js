import React, { Fragment } from 'react'
import { css } from '@emotion/core'
import { useQuery } from '@apollo/client'
import GET_TABLE_DATA from '../gql/transactions.gql'
import { TxTable } from '../components/transactions/TxTable'
import CircularProgress from '@material-ui/core/CircularProgress'

export function Transactions () {
  const [data, setData] = React.useState({ transactions: [] })
  const { loading, error } = useQuery(GET_TABLE_DATA, {
    onCompleted: fetched => setData(fetched)
  })

  if (loading) {
    return (
      <div css={loadingStyle}>
        <CircularProgress color='secondary' />
      </div>
    )
  }

  if (error) {
    return <Fragment>¯\_(ツ)_/¯</Fragment>
  }

  return (
    <Fragment>
      <TxTable data={data} setData={setData} />
    </Fragment>
  )
}

const loadingStyle = css`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 50vh;
`
