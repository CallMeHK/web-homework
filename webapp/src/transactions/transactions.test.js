jest.mock('../components/transactions/TxTable', () => {
  return {
    TxTable: props => <div>table</div>
  }
})

import React from 'react'
import renderer from 'react-test-renderer'
import { waitFor, fireEvent, render } from '@testing-library/react'
import { Transactions } from './transactions'
import * as Apollo from '@apollo/client'

import MutationObserver from 'mutation-observer'
global.MutationObserver = MutationObserver

describe('transactions', () => {
  let useQueryStub
  beforeEach(() => {
    useQueryStub = jest.spyOn(Apollo, 'useQuery')
  })
  describe('Transactions', () => {
    describe('Loading', () => {
      it('should render loading experience', () => {
        useQueryStub.mockImplementation(() => ({ loading: true }))
        const tree = renderer.create(<Transactions />).toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
    describe('Error', () => {
      it('should render error experience', () => {
        useQueryStub.mockImplementation(() => ({ error: true }))
        const tree = renderer.create(<Transactions />).toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('Experience loaded', () => {
      it('should render error experience', async () => {
        useQueryStub.mockImplementationOnce((_, { onCompleted }) => {
          onCompleted({ merchants: [], users: [], transactions: [] })
          return {}
        })
        useQueryStub.mockImplementation(() => ({}))
        const wrapper = render(<Transactions />)

        expect(await wrapper.findByText('table')).toBeInTheDocument()
      })
    })
  })
})
