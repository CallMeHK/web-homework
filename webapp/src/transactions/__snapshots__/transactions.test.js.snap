// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions Transactions Error should render error experience 1`] = `"¯\\\\_(ツ)_/¯"`;

exports[`transactions Transactions Loading should render loading experience 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 50vh;
}

<div
  className="emotion-0"
>
  <div
    className="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
